File: epoxy-sample/src/main/java/com/airbnb/epoxy/sample/SampleController.java
package com.airbnb.epoxy.sample;

import com.airbnb.epoxy.AutoModel;
import com.airbnb.epoxy.TypedEpoxyController;
import com.airbnb.epoxy.sample.models.CarouselModelGroup;
import com.airbnb.epoxy.sample.views.HeaderViewModel_;

import java.util.List;

import static com.airbnb.epoxy.EpoxyAsyncUtil.getAsyncBackgroundHandler;

public class SampleController extends TypedEpoxyController<List<CarouselData>> {
  public interface AdapterCallbacks {
    void onAddCarouselClicked();
    void onClearCarouselsClicked();
    void onShuffleCarouselsClicked();
    void onChangeAllColorsClicked();
    void onAddColorToCarouselClicked(CarouselData carousel);
    void onClearCarouselClicked(CarouselData carousel);
    void onShuffleCarouselColorsClicked(CarouselData carousel);
    void onChangeCarouselColorsClicked(CarouselData carousel);
    void onColorClicked(CarouselData carousel, int colorPosition);
  }

  @AutoModel HeaderViewModel_ header;
  @AutoModel ButtonBindingModel_ addButton;
  @AutoModel ButtonBindingModel_ clearButton;
  @AutoModel ButtonBindingModel_ shuffleButton;
  @AutoModel ButtonBindingModel_ changeColorsButton;

  private final AdapterCallbacks callbacks;

  SampleController(AdapterCallbacks callbacks) {
    // Demonstrating how model building and diffing can be done in the background.
    // You can control them separately by passing in separate handler, as shown below.
    super(getAsyncBackgroundHandler(), getAsyncBackgroundHandler());
//    super(new Handler(), BACKGROUND_HANDLER);
//    super(BACKGROUND_HANDLER, new Handler());

    this.callbacks = callbacks;
    setDebugLoggingEnabled(true);
  }

  @Override
  protected void buildModels(List<CarouselData> carousels) {
    header
        .title(R.string.epoxy)
        .caption(R.string.header_subtitle);
    // "addTo" is not needed since implicit adding is enabled
    // (https://github.com/airbnb/epoxy/wiki/Epoxy-Controller#implicit-adding)

    addButton
        .textRes(R.string.button_add)
        .clickListener((model, parentView, clickedView, position) -> {
          callbacks.onAddCarouselClicked();
        });

    clearButton
        .textRes(R.string.button_clear)
        .clickListener(v -> callbacks.onClearCarouselsClicked())
        .addIf(carousels.size() > 0, this);

    shuffleButton
        .textRes(R.string.button_shuffle)
        .clickListener(v -> callbacks.onShuffleCarouselsClicked())
        .addIf(carousels.size() > 1, this);

    changeColorsButton
        .textRes(R.string.button_change)
        .clickListener(v -> callbacks.onChangeAllColorsClicked())
        .addIf(carousels.size() > 0, this);

    for (int i = 0; i < carousels.size(); i++) {
      CarouselData carousel = carousels.get(i);
      add(new CarouselModelGroup(carousel, callbacks));
    }
  }

  @Override
  protected void onExceptionSwallowed(RuntimeException exception) {
    // Best practice is to throw in debug so you are aware of any issues that Epoxy notices.
    // Otherwise Epoxy does its best to swallow these exceptions and continue gracefully
    throw exception;
  }
}

File: epoxy-sample/src/main/java/com/airbnb/epoxy/sample/SampleController.kt
package com.airbnb.epoxy.sample

import android.view.View
import com.airbnb.epoxy.AutoModel
import com.airbnb.epoxy.DataBindingEpoxyModel.DataBindingHolder
import com.airbnb.epoxy.EpoxyAsyncUtil
import com.airbnb.epoxy.TypedEpoxyController
import com.airbnb.epoxy.sample.models.CarouselModelGroup
import com.airbnb.epoxy.sample.views.HeaderViewModel_

class SampleController(private val callbacks: AdapterCallbacks) :
    TypedEpoxyController<List<CarouselData>>(
        EpoxyAsyncUtil.getAsyncBackgroundHandler(),
        EpoxyAsyncUtil.getAsyncBackgroundHandler()
    ) {
    interface AdapterCallbacks {
        fun onAddCarouselClicked()
        fun onClearCarouselsClicked()
        fun onShuffleCarouselsClicked()
        fun onChangeAllColorsClicked()
        fun onAddColorToCarouselClicked(carousel: CarouselData?)
        fun onClearCarouselClicked(carousel: CarouselData?)
        fun onShuffleCarouselColorsClicked(carousel: CarouselData?)
        fun onChangeCarouselColorsClicked(carousel: CarouselData?)
        fun onColorClicked(carousel: CarouselData?, colorPosition: Int)
    }

    @AutoModel
    lateinit var header: HeaderViewModel_

    @AutoModel
    lateinit var addButton: ButtonBindingModel_

    @AutoModel
    lateinit var clearButton: ButtonBindingModel_

    @AutoModel
    lateinit var shuffleButton: ButtonBindingModel_

    @AutoModel
    lateinit var changeColorsButton: ButtonBindingModel_

    override fun buildModels(carousels: List<CarouselData>) {
        header
            .title(R.string.epoxy)
            .caption(R.string.header_subtitle)
        // "addTo" is not needed since implicit adding is enabled
        // (https://github.com/airbnb/epoxy/wiki/Epoxy-Controller#implicit-adding)
        addButton
            .textRes(R.string.button_add)
            .clickListener { model: ButtonBindingModel_?, parentView: DataBindingHolder?, clickedView: View?, position: Int -> callbacks.onAddCarouselClicked() }
        clearButton
            .textRes(R.string.button_clear)
            .clickListener { v: View? -> callbacks.onClearCarouselsClicked() }
            .addIf(carousels.size > 0, this)
        shuffleButton
            .textRes(R.string.button_shuffle)
            .clickListener { v: View? -> callbacks.onShuffleCarouselsClicked() }
            .addIf(carousels.size > 1, this)
        changeColorsButton
            .textRes(R.string.button_change)
            .clickListener { v: View? -> callbacks.onChangeAllColorsClicked() }
            .addIf(carousels.size > 0, this)
        for (i in carousels.indices) {
            val carousel = carousels[i]
            add(CarouselModelGroup(carousel, callbacks))
        }
    }

    override fun onExceptionSwallowed(exception: RuntimeException) {
        // Best practice is to throw in debug so you are aware of any issues that Epoxy notices.
        // Otherwise Epoxy does its best to swallow these exceptions and continue gracefully
        throw exception
    }

    init {
        // Demonstrating how model building and diffing can be done in the background.
        // You can control them separately by passing in separate handler, as shown below.
        //    super(new Handler(), BACKGROUND_HANDLER);
//    super(BACKGROUND_HANDLER, new Handler());
        isDebugLoggingEnabled = true
    }
}
